TII-188 TII LTI integration

bgarciaentornos opened the pull request:
Changes needed on the assignments, basiclti and content-review tools in order to support the new TII LTI integration.

bbailla2 commented on assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java:
```diff
@@ -2537,6 +2550,32 @@ else if (submittedTime == null) /*grading non-submission*/
+			catch (Exception e)
+			{
+				M_log.warn(" commitEditFromCallback(): closed AssignmentSubmissionEdit assignment submission id=" + submission.getId() + e.getMessage());
+			}
```
Thrown exception is wasteful if we're going to catch it immediately. I'd remove the try/catch and just log

bbailla2 commented on assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java:
```diff
@@ -6004,6 +6043,231 @@ public boolean willArchiveMerge()
+					throw e;
+				}
+				catch (Exception ignore)
+				{
```
Should log something here; swallowing would make it very difficult to troubleshoot should something go wrong

bbailla2 commented on assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java:
```diff
@@ -6015,9 +6279,68 @@ public void handleAccess(HttpServletRequest req, HttpServletResponse res, Refere
+							newsession.setActive();
+							SessionManager.setCurrentSession(newsession);
+							newsession.setUserId("admin");
+							newsession.setUserEid("admin");
```
If you need to temporarily elevate permissions, use the "yesMan" SecurityAdvisor pattern; should never switch the session to "admin". The SecurityAdvisor is tied only to the current thread, so it's much much safer in terms of multi-tabbing or anything with concurrency and so forth.

teetoppz28 replied:
You could also leverage the DevelopHelperService to set a user temporarily. Such as:

developerHelperService.setCurrentUser(DeveloperHelperService.ADMIN_USER_REF);
... do your stuff ...
developerHelperService.restoreCurrentUser();

bbailla2 replied:
Consider pseudocode like:
currentUser = getCurrentUser(); setCurrentUser(admin); do stuff; restoreCurrentUser();
I've encountered this issue before - simply by double clicking, you spawn two threads:
Thread 1:
currentUser = ""; setCurrentUser(admin); do stuff;
Thread 2:
currentUser = admin; setCurrentUser(admin); do stuff;
thread 1:
restoreCurrentUser() // becomes ""
thread 2:
restoreCurrentUser() // becomes "admin"

Then your user remains an administrator, and very very bad things happen. SecurityAdvisor is tied to the current thread and I highly advise using it instead.

bjones86 replied:
I'm in agreement with @bbailla2 on this one, we should always be using SecurityAdvisors in scenarios like this. We learned the hard way; "very very bad things happen."

bbailla2 commented on assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:
```diff
@@ -1003,6 +1023,16 @@ public String buildMainPanelContext(VelocityPortlet portlet, Context context, Ru
+				isDirectAccess = true;
+			} else {
+				isDirectAccess = false;
+			}
```
Cut this down to:
allowLTIReviewService = contentReviewSiteAdvisor.siteCanUseLTIReviewService(s);
isDirectAccess = conrentReviewService.isDirectAccess(s);

bbailla2 commented on assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:
```diff
@@ -1560,6 +1598,23 @@ protected String build_student_view_submission_context(VelocityPortlet portlet,
+				context.put("ltiLink", ltiLink);
+				int factor = AssignmentService.getScaleFactor();
+				int dec = (int)Math.log10(factor);
+				int maxPointsInt = assignment.getContent().getMaxGradePoint() / dec;
```
Previous 3 lines are repeated a few times throughout the patch, should be refactored into a method

bbailla2 commented on assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:
```diff
@@ -6783,7 +6908,9 @@ public SecurityAdvice isAllowed(String userId, String function, String reference
-			contentReviewService.queueContent(contentUserId, siteId, edit.getAssignment().getReference(), Arrays.asList(attachment));
+			if(!isResubmission){
+				contentReviewService.queueContent(null, null, edit.getAssignment().getReference(), Arrays.asList(attachment), edit.getId(), isResubmission);
+			}
```
Are we preventing resubmissions from being checked with the content-review service?

bgarciaentornos replied:
This is a bug, the if line should be removed.

bbailla2 commented on assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:
```diff
@@ -7264,6 +7461,12 @@ else if (sameAssignmentTitleInContext(assignmentRef, title, (String) state.getAt
+		//set whether students can view the review service grades
+		r = params.getString(NEW_ASSIGNMENT_ALLOW_STUDENT_VIEW_EXTERNAL_GRADE);
+		if (r == null) b = Boolean.FALSE.toString();
+		else b = Boolean.TRUE.toString();
```
This class is huge, so we should take every opportunity to condense things here. I'd change these if.. else checks to one liners like:
b = String.valueOf(r != null);

bbailla2 commented on assignment/assignment-tool/tool/src/webapp/vm/assignment/chef_assignments_instructor_grading_submission.vm:
```diff
@@ -234,10 +234,10 @@
 										<span class="reportIcon">
 										#if ($reviewResult.getReviewReport().equals("Error"))
-											<img src="/library/skin/images/error.png" alt="$reviewResult.getReviewError()" title="$reviewResult.getReviewError()">
+											<i alt="$reviewResult.getReviewError()" title="$reviewResult.getReviewError()" class="fa fa-exclamation-triangle errorStyle"></i>
```
Stick to `<img>` or `<span>` or such; the `<i>` tag should be used strictly for text in italics

bbailla2 commented on basiclti/basiclti-common/src/java/org/sakaiproject/basiclti/util/SakaiBLTIUtil.java:
```diff
@@ -935,10 +935,83 @@ public static void addGlobalData(Site site, Properties props, Properties custom,
+		
+		//substitute resources params and add the rest
+		if(custom.getProperty(BasicLTIConstants.RESOURCE_LINK_ID) != null){//TODO if allowReviewService prop y future prop for TII
+			addTIIproperties(ltiProps, custom);
```
This project should be solely focused on LTI without code specific to any individual LTI integrations; addTIIproperties should be moved to the Turnitin content-review implementation

jonespm replied:
I don't see the point of this code anyway. All of this data is contained in the launch. If turnitin can't process it for some reason that sounds more like their problem. But if you can't work around them, then this code should be a lot more generic. (Maybe a list of custom properties to "promote" that could be set via configuration. I wouldn't merge with this very specific code in here.

bgarciaentornos replied:
We need to override the resource id, title and description, so we can use the assignment values instead of the site values. That can't be done for each LTI launch within the CR code, that's why it appears here (as out of place as it seems). It would be cool to create some generic methods for special LTI launches like this one, that wouldn't have to rely on TII specific code, but we were in a rush and this was a quick workaround that didn't affect the other LTI calls.

ern commented:
FYI while this patch is specific for 11.x branch a new tii implementation will be added to master, reflecting these changes.

bbailla2 commented on assignment/assignment-api/api/src/java/org/sakaiproject/assignment/api/ContentReviewResult.java:
```diff
@@ -33,6 +33,16 @@
+	private String reviewIconColor;
+	
+	/**
+	 *  Indicates whether the external grade for this submission is different than the assignments one
```
I believe this stores the external grade itself

bbailla2 commented on assignment/assignment-bundles/resources/assignment.properties:
```diff
@@ -898,6 +902,7 @@ content_review.pending.info = This attachment has been submitted and is pending
 content_review.error = An unknown error occurred. The originality review for this attachment is not available.
 content_review.error.createAssignment=An error with {0} has occurred while creating this assignment. {1} has saved the assignment in draft mode. Please try posting this assignment again later. {2}
 content_review.note=<div><br /><em>NOTE:</em><ul><li>When submitting attachments, students should only use these file types: {0}.</li> <li>Students should always save files with the appropriate extension.</li> </ul></div>
+content_review.studentNote = Only use file types: Microsoft Word (.doc, .docx), PowerPoint (.ppt, .pptx), PostScript (.ps), PDF (.pdf), HTML (.html), OpenOffice (ODT), Hangul (HWP), rich or plain text (.rtf, .txt).</li><li>Always include file extension.</li><li>Maximum file size is 20 Mb. The document must contain more than 20 words.</li></ol>
```
Should paramaterize the file type list passing in AssignmentAction.getContentReviewAcceptedFileTypesMessage()

bbailla2 commented on assignment/assignment-bundles/resources/assignment.properties:
```diff
@@ -1008,3 +1013,21 @@ grades.late=Late submission
+review.user.lastname=The content review service requires your surname to be set. Your submission cannot be reviewed by {0} until you update your details.
+review.instructor.fields=The content review service requires your first name, surname and email must to be set. If you wish to route submissions through {0} then you need to update your details.
+review.conflicting.popup=This item has been graded in both Turnitin and in the Assignments Tool. The mark set in the Assignments Tool is the one that is displayed here.
+allowResubmission.review.warning=Any submissions made after the Assignment Due Date will not be sent to {0}
```
A few grammar revisions could be made in some of these. Not sure if regular users know what "content review service" means, I'd just put {0} twice (Ie. 'Turnitin').

bbailla2 commented on assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java:
```diff
@@ -6004,6 +6043,231 @@ public boolean willArchiveMerge()
+                len = -1;
+            }
+            if (len < buffer.length)
+                break;
```
I'm guessing this is also in BaseContentService, but this 'if... break' is redundant; when len < buffer.length, the while's evaluation ((bytesToRead > 0) && (len >= buffer.length)) will be false.

ern commented:
The changes to assignments are to specific to Turnitin I think you need to push this back into the Turnitin implementation and make it generic in Assignments. Would like to prevent Assignments knowing more than it needs to about which integration is actually in use. I am not saying that there won't be some things that need to know like the UI elements but we should keep things as generic as possible...

bbailla2 commented:
I mostly agree with you Earl, but there's quite a bit in here that necessitates changes to the behavior of the assignments tool: LTI direct linking, the TII callback to complete assignment creation, the introduction of external grades, etc.

bbailla2 commented on assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java:
```diff
@@ -10463,19 +10821,31 @@ public String getReviewReport() {
+							} else {
+								return contentReviewService.getReviewReportStudent(contentId, getAssignment().getReference(), UserDirectoryService.getCurrentUser().getId());
+							}
+						}
```
Couple of these blocks; they read simpler as
if (siteCanUseLTIReviewService)...
else if (allowGradeSubmission....
else...

bbailla2 commented on assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java:
```diff
@@ -10518,13 +10896,13 @@ private String getReviewReport(ContentResource cr)
 				Reference ref = (Reference)m_submittedAttachments.get(i);
 				ContentResource contentResource = (ContentResource)ref.getEntity();
-				if (contentReviewService.isAcceptableContent(contentResource)) {
+				if (contentReviewService.isAcceptableSize(contentResource) && (allowAnyFile || contentReviewService.isAcceptableContent(contentResource))){
```
The intention of contentReviewService.isAcceptableContent(ContentResource) was to handle all of these checks in one spot. isAcceptableSize and allowAnyFile should be refactored into CRS.isAcceptableContent()

bbailla2 commented on assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java:
```diff
@@ -12597,14 +12998,50 @@ public void postAttachment(List attachments){
+				e.printStackTrace();
+			}
+
+		}
```
Refactoring opportunity - this method is identical to postAttachment(List attachments) except the queueContent vs queueResubContent.

bbailla2 commented on assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:
```diff
@@ -196,6 +198,14 @@
 	private static final Boolean allowReviewService = ServerConfigurationService.getBoolean("assignment.useContentReview", false);
 	private static final Boolean allowPeerAssessment = ServerConfigurationService.getBoolean("assignment.usePeerAssessment", true);
+	private static Boolean allowLTIReviewService = false;
+	private static Boolean isDirectAccess = false;
```
These static variables are referenced across multiple user sessions, so this is likely to cause concurrency issues. For instance, two users working in sites with opposite ContentReviewSiteAdvisor settings will flip these on eachother.

Now, AssignmentAction is a servlet-class, and so I believe it's somewhat of a singleton, so simply removing the static won't fix this; the ContentReviewSiteAdvisor will have to be consulted for every usage these.

bbailla2 commented on assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:
```diff
@@ -9521,12 +9735,18 @@ public boolean createTIIAssignment(AssignmentContentEdit assign, String assignme
         opts.put("submit_papers_to", assign.getSubmitReviewRepo());
-        opts.put("report_gen_speed", assign.getGenerateOriginalityReport());
+        String originalityReportVal = assign.getGenerateOriginalityReport();
+        if(originalityReportVal.equals(NEW_ASSIGNMENT_REVIEW_SERVICE_REPORT_IMMEDIATELY) && state.getAttribute(AssignmentSubmission.ALLOW_RESUBMIT_NUMBER) != null){
```
Yoda expression recommended

bbailla2 commented on assignment/assignment-tool/tool/src/java/org/sakaiproject/assignment/tool/AssignmentAction.java:
```diff
@@ -17028,6 +17327,11 @@ public SecurityAdvice isAllowed(String userId, String function, String reference
+						if(allowReviewService && !contentReviewService.isAcceptableContent(attachment)) {
+							addAlert(state, rb.getFormattedMessage("cr.notprocess.warning"));
+						} else if(!contentReviewService.isAcceptableSize(attachment)) {
+							addAlert(state, rb.getFormattedMessage("cr.size.warning"));
```
"else if" executes regardless of 'allowReviewService'

bjones86 commented:
This has been open for months now and needs to be rebased in order to be merged. @bgarciaentornos are you planning on addressing the issues/suggestions brought up by the reviewers?

I'd like this to get in sooner than later so that other people can do some work on top of this.

bgarciaentornos commented:
Well, long story short: this was initially planned for 11 but it was
decided Earle would do some work to mix it on master/trunk with his CR
reorganization and then close the PR. Plus right now there's one 11 version
made by Charles Hedrick and that other Earle version (although I believe he
finally didn't migrate this part). I guess this PR can be closed, but I'm
not sure how should the changes be addressed considering all these other
existing versions. I'd say none of the changes imply behaviour change bar
the ones on the LTI tool and some on the assignments, suggested by Earle,
Both of those pieces of code are hard to take elsewhere though.

2016-06-28 18:26 GMT+02:00 Brian Jones <notifications@github.com>:

> This has been open for months now and needs to be rebased in order to be
> merged. @bgarciaentornos <https://github.com/bgarciaentornos> are you
> planning on addressing the issues/suggestions brought up by the reviewers?
>
> I'd like this to get in sooner than later so that other people can do some
> work on top of this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sakaiproject/sakai/pull/2202#issuecomment-229103614>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AK6IltOIMpSw-3dBsAX5f0SRVO19bvofks5qQUs9gaJpZM4IFIaz>
> .
>

bgarciaentornos commented:
But not exactly. I think some of that work has already started taking place on the contrib branch, forked from Chuck's repo (https://github.com/clhedrick/turnitin). He even modified some of the assignment UI, but I don't really know where's that code. So imo the next PR (to the master version as accorded, then backport) should probably come from that code, taking care of all the suggestions made here (thanks to all!) and considering Earle's modifications to the CR that concern these tools. Let's ask them about this: @clhedrick  @ern

clhedrick commented:
My code is more or less ready for production I think. I made major changes to the UI in assignment. It hadn’t been styled at all.

We still need to do something about showing the result from an assignment create, but I’m holding action on that because it sounds like TII is adding a new feature that will let us do a better job.

On Jun 29, 2016, at 11:02 AM, bgarciaentornos <notifications@github.com<mailto:notifications@github.com>> wrote:


But not exactly. I think some of that work has already started taking place on the contrib branch, forked from Chuck's repo (https://github.com/clhedrick/turnitin). He even modified some of the assignment UI, but I don't really know where's that code. So imo the next PR (to the master version as accorded, then backport) should probably come from that code, taking care of all the suggestions made here (thanks to all!) and considering Earle's modifications to the CR that concern these tools. Let's ask them about this: @clhedrick<https://github.com/clhedrick> @ern<https://github.com/ern>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/sakaiproject/sakai/pull/2202#issuecomment-229384536>, or mute the thread<https://github.com/notifications/unsubscribe/AB0aCEFHJp863sslq5Vaj9cfo8HgI9m3ks5qQoj5gaJpZM4IFIaz>.

bjones86 commented:
So does this mean that all the code from this PR for assigments, basicLTI and content-review are already in master/11.x?

What about the turnitin code here? Do we already have that in sakaicontrib/turnitin?

clhedrick commented:
Yes. there’s a copy of contentreview-impl and a diff for everything else.

On Jun 29, 2016, at 11:14 AM, Brian Jones <notifications@github.com<mailto:notifications@github.com>> wrote:


So does this mean that all the code from this PR for assigments, basicLTI and content-review are already in master/11.x?

What about the turnitin code here? Do we already have that in sakaicontrib/turnitin?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/sakaiproject/sakai/pull/2202#issuecomment-229388391>, or mute the thread<https://github.com/notifications/unsubscribe/AB0aCEAt0awcZkUECIvRRTtQ-PRdr6uTks5qQovLgaJpZM4IFIaz>.

bjones86 commented:
Okay, I'm still slightly confused here.

"Yes. there’s a copy of contentreview-impl and a diff for everything else."

Are you saying that all the code from this PR has already been merged into assignments, basicLTI and content-review in master/11?

"there’s a copy of contentreview-impl" are you referring to sakaicontrib/turnitin here? And does that now contain all relevant code from this PR?

"and a diff for everything else." A diff doesn't really help us, we need the code in the code base for other people to work on top of it. What is this diff you're referring to?

ern commented:
This PR will more than likely be closed here.
And I think @buckett is working on cleaning up some of this up so I am going to reassign to him to close. Once the new turnitin is in a good state I will look at patching this back into master.

bjones86 commented:
Ya, I think this can be closed. We have sakaicontrib/turnitin and sakaicontrib/sakai-turnitin for this purpose. The commits from sakaicontrib/sakai-turnitin can be merged to the main sakaiproject/sakai code base when everything is tested and all the different parties are happy with how it's implemented and works.

bjones86 commented:
Closing this. We agreed on the Core Team call this week that all TII LTI work will be done in sakaicontrib/sakai-turnitin and sakaicontrib/turnitin. When everything is there, working properly, and everyone is happy with it, the commits from sakaicontrib/sakai-turnitin can be merged back to sakaiproject/sakai.
